backend/guides/code.md

```diff
@@ -0,0 +1,159 @@
+# Coding
+
+## Source code
+
+### Naming above all
+
+The source code you write is intended for only two audiences
+
+- other people,
+- compiler/interpreter.
+
+The compiler doesn't care about the naming, doesn't have to understand the semantics, therefore the readability and understandability are the most important aspects above all.
+
+It's exactly like when two people from different countries trying to express themselves with language the other doesn't understand. Or even worse, they speak the same language and each of them are using different naming for the same things.
+
+When you call something 'banana', it should be banana and not an orange.
+
+### Learn to love old code
+
+If it works, there is no need to refactor it. Refactoring a code that is old but checks out can be done over and over and forever.
+
+Do you need to add functionality? Compose the code.
+
+### Minimal, conscious, self-documenting
+
+1. Less code is oftentimes better, than more code, if it does not sacrifice readability or comprehension. Smaller codebase is more maintainable and shorter snippets are faster to review and understand.
+2. Avoid using one-liners if it hurts the readability
+
+```js
+// BAD
+const articles = getArticles()
+const filteredArticles = []
+for (const article of articles) {
+    if (isValid(article)) {
+        filteredArticles.push(article)
+    }
+}
+// GOOD
+const filteredArticles = getArticles().filter(isValid)
+```
+
+As seen in the example, limiting local variables and imperative constructs usually leads to more readable code, that does not need comments and is easier to follow.
+
+This practice is sometimes referred to as [💋 KISS](https://en.wikipedia.org/wiki/KISS_principle)
+
+## Prefer libraries
+
+Always prefer libraries above your code - you transfer the responsibility to the libraries, that usually are well tested
+
+Get to know our libs!
+
+- Logging. https://github.com/AckeeCZ/cosmas
+- Configuration. https://github.com/AckeeCZ/configuru
+- User notifications. https://github.com/AckeeCZ/enmail
+- Message queues. https://github.com/AckeeCZ/fuqu
+- Utilities. https://lodash.com/ (or FP variant), https://ramdajs.com/
+- MySQL database. https://github.com/AckeeCZ/databless
+- Authentication. https://github.com/AckeeCZ/authist
+- HTTP Server. https://github.com/AckeeCZ/unicore
+- ACL. https://github.com/AckeeCZ/axesor
+
+## Typescript
+
+1. Don't `export` if you don't need the thing exported, just because you think, you might need it in next commit or distant future.
```

Comment:
Link YAGNI principle

Reply:
I have objection to this rule if it includes types. From a DX view, if you ever export a function, all of its types should be exported too. So give types an exception? :D 

E.g. if you export the following function, you should export `Bar` and `Baz`, because there is a fair chance, you will want to create `bar: Bar` in advance and then pass it it. And nothing makes the code less readable than using `Parameters<typeof foo>[0]`
```ts
foo(bar: Bar): Baz
```

Reply:
I don't think types should have a general exception.

1. You don't have types like `Bar` and `Baz`, but `User`, `Descriptor`, etc. That collides with lot of things so people try to prefix that which leads to unnecessary complicated interface names that are (possibly) only used locally.
2. This might not be possible with the use of generics, because you might need to guarantee a binding between arguments.
3. What is the complexity you should create a labeled exported type? `string` (no, primitive), `string | undefined` (no, primitive optional parameter), `string | null` (possibly?), `Array<string>` (probably not?), `[string, number]` (probably not?), `[number: min, number: max]` (possibly with the labels?)

I agree that sometimes it is helpful, but I think we should leave that to the devs, because there is no clear way to say when you should export argument types and when not.

Reply:
From my experience most of the object options.

Take Firebase as an example of input credentials setup object for app https://github.com/firebase/firebase-admin-node/blob/9dd4040888d7d6ac7c2970bad26e7646d6a3f384/src/firebase-namespace-api.ts#L266. Programmers then tend to create their interfaces (usually incorrect, causing problems later) or using `any`.

The similar goes for output object interfaces (`() -> Out)`, which you need to use in order to create your function to process the results. This has the same consequences. Which sounds like ReturnType solves the problem, until the return type is Promise<Out> and you have to unwrap it :/.

General namings are and always have been solved by scoping - you can always use `somemodule.Foo`. This way firebase, but not only firebase, can have easily options called Options, Credentials etc.

Reply:
This is fine - https://github.com/AckeeCZ/styleguide/pull/12/commits/f9220e650916842a43091caa96c5079d17f568ca#diff-cc38b01a70e42c94b7d20874addfe15a644010f809ebc53777a9ed10eae950b3R69

Resolving discussion.